java/core/src/main/java/com/google/protobuf/CodedInputStream.java

```diff
@@ -1246,18 +1256,26 @@ public byte readRawByte() throws IOException {
 
     @Override
     public byte[] readRawBytes(final int length) throws IOException {
+      byte[] bytes = EMPTY_BYTE_ARRAY;
       if (length > 0 && length <= (limit - pos)) {
+        bytes = new byte[length];
+      }
+      readRawBytes(bytes, 0, length);
+      return bytes;
+    }
+
+    @Override
+    public void readRawBytes(final byte[] dst, final int offset, final int length)
+        throws IOException {
+      if (length >= 0 && length <= (limit - pos)) {
         final int tempPos = pos;
         pos += length;
-        return Arrays.copyOfRange(buffer, tempPos, pos);
+        System.arraycopy(buffer, tempPos, dst, offset, length);
```

Comment:
This throws 3 exceptions we should catch and rethrow as a runtime InvalidProtocolBufferException or IOException, whichever makes the most sense: https://github.com/AdoptOpenJDK/openjdk-jdk11/blob/master/src/java.base/share/classes/java/lang/System.java#L522